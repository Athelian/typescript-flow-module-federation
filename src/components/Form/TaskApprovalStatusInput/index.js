// @flow
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import Icon from 'components/Icon';
import UserAvatar from 'components/UserAvatar';
import FormattedDate from 'components/FormattedDate';
import { type UserAvatarType } from 'types';
import { getByPathWithDefault } from 'utils/fp';
import {
  TaskApprovalStatusInputWrapperStyle,
  UserAvatarWrapperStyle,
  DeactivateButtonStyle,
  StatusLabelWrapperStyle,
  StatusLabelStyle,
} from './style';

type OptionalProps = {
  width: string,
  approval?: ?{
    approvedBy: UserAvatarType,
    approvedAt: string,
  },
  rejection?: ?{
    rejectedBy: UserAvatarType,
    rejectedAt: string,
  },
  showUser: boolean,
  onClickUser: () => void,
  showDate: boolean,
  editable: boolean,
};

type Props = OptionalProps & {};

const defaultProps = {
  width: '200px',
  showUser: false,
  onClickUser: () => {},
  showDate: false,
  editable: false,
};

const TaskApprovalStatusInput = ({
  width,
  approval,
  rejection,
  showUser,
  onClickUser,
  showDate,
  editable,
}: Props) => {
  const parsedUser = {
    firstName: approval
      ? getByPathWithDefault('', 'approvedBy.firstName', approval)
      : getByPathWithDefault('', 'rejectedBy.firstName', rejection),
    lastName: approval
      ? getByPathWithDefault('', 'approvedBy.lastName', approval)
      : getByPathWithDefault('', 'rejectedBy.lastName', rejection),
  };
  const parsedDate = approval
    ? getByPathWithDefault('', 'approvedAt', approval)
    : getByPathWithDefault('', 'approvedAt', rejection);

  return (
    <div className={TaskApprovalStatusInputWrapperStyle(!!approval, width)}>
      {showUser && (
        <div className={UserAvatarWrapperStyle}>
          <UserAvatar firstName={parsedUser.firstName} lastName={parsedUser.lastName} />
          {editable && (
            <button
              className={DeactivateButtonStyle}
              onClick={evt => {
                evt.stopPropagation();
                onClickUser();
              }}
              type="button"
            >
              <Icon icon="CLEAR" />
            </button>
          )}
        </div>
      )}

      <div className={StatusLabelWrapperStyle}>
        <div className={StatusLabelStyle}>
          {approval ? (
            <FormattedMessage id="components.inputs.approved" defaultMessage="APPROVED" />
          ) : (
            <FormattedMessage id="components.inputs.rejected" defaultMessage="REJECTED" />
          )}
        </div>

        {showDate && (
          <div className={StatusLabelStyle}>
            <FormattedDate value={parsedDate} />
          </div>
        )}
      </div>

      <Icon icon={approval ? 'CHECKED' : 'CANCEL'} />
    </div>
  );
};

TaskApprovalStatusInput.defaultProps = defaultProps;

export default TaskApprovalStatusInput;
