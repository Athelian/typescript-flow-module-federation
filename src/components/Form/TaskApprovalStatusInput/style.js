// @flow
import { css } from 'react-emotion';
import { presets, borderRadiuses, fontSizes, colors, shadows } from 'styles/common';

export const TaskApprovalStatusInputWrapperStyle = (
  isApproved: boolean,
  width: string
): string => css`
  ${presets.BUTTON};
  ${borderRadiuses.BUTTON};
  height: 40px;
  width: ${width};
  cursor: inherit;
  padding: 3px;
  ${isApproved
    ? `
    background-color: ${colors.BLUE};
    border: 2px solid ${colors.WHITE};
    color: ${colors.WHITE};
  `
    : `
    background-color: ${colors.RED};
    border: 2px solid ${colors.RED};
    color: ${colors.WHITE};
  `};
`;

export const UserAvatarWrapperStyle: string = css`
  position: relative;
  &:hover {
    & > button {
      opacity: 1;
    }
  }
`;

export const DeactivateButtonStyle: string = css`
  position: absolute;
  top: 0;
  ${presets.BUTTON};
  opacity: 0;
  ${borderRadiuses.CIRCLE};
  background-color: ${colors.WHITE};
  color: ${colors.GRAY_LIGHT};
  height: 30px;
  width: 30px;
  flex-shrink: 0;
  cursor: inherit;
  ${shadows.DROPDOWN};
  &:hover,
  :focus {
    color: ${colors.RED};
  }
`;

export const StatusLabelWrapperStyle: string = css`
  flex: 1;
`;

export const StatusLabelStyle: string = css`
  ${presets.ELLIPSIS};
  ${fontSizes.SMALL};
  letter-spacing: 2px;
  user-select: none;
  text-transform: uppercase;
  text-align: center;
`;
