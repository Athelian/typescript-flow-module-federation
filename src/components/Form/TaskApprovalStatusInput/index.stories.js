/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { ObjectValue } from 'react-values';
import { TaskApprovalStatusInput, ToggleInput, Label } from 'components/Form';

storiesOf('Form/Inputs/Task Approval Status Input', module).add(
  'Task Approval Status Input',
  () => (
    <>
      <ObjectValue
        defaultValue={{
          showUser: true,
          editable: true,
        }}
      >
        {({ value, set }) => {
          const { editable, showUser } = value;

          return (
            <>
              <TaskApprovalStatusInput
                {...value}
                approval={{
                  approvedBy: { id: 1, firstName: 'Kevin', lastName: 'Nguyen' },
                  approvedAt: new Date(),
                }}
                rejection={null}
              />
              <ToggleInput toggled={showUser} onToggle={() => set('showUser', !showUser)}>
                <Label>SHOW USER</Label>
              </ToggleInput>
              <ToggleInput toggled={editable} onToggle={() => set('editable', !editable)}>
                <Label>EDITABLE</Label>
              </ToggleInput>
            </>
          );
        }}
      </ObjectValue>

      <ObjectValue
        defaultValue={{
          showUser: true,
          editable: true,
        }}
      >
        {({ value, set }) => {
          const { editable, showUser } = value;

          return (
            <>
              <TaskApprovalStatusInput
                {...value}
                approval={null}
                rejection={{
                  rejectedBy: { id: 1, firstName: 'Kevin', lastName: 'Nguyen' },
                  rejectedAt: new Date(),
                }}
              />
              <ToggleInput toggled={showUser} onToggle={() => set('showUser', !showUser)}>
                <Label>SHOW USER</Label>
              </ToggleInput>
              <ToggleInput toggled={editable} onToggle={() => set('editable', !editable)}>
                <Label>EDITABLE</Label>
              </ToggleInput>
            </>
          );
        }}
      </ObjectValue>
    </>
  )
);
