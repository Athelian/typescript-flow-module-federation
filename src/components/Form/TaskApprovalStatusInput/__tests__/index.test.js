import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { IntlProvider } from 'react-intl';
import TaskApprovalStatusInput from '../index';

describe('TaskApprovalStatusInput', () => {
  it('approved, showUser, showDate', () => {
    const approval = {
      approvedBy: {
        firstName: 'TJ',
        lastName: 'Jiang',
      },
      approvedAt: '2019-04-24',
    };
    render(
      <IntlProvider locale="en">
        <TaskApprovalStatusInput approval={approval} showUser showDate />
      </IntlProvider>
    );

    expect(screen).toMatchSnapshot();
    expect(screen.getByText(/APPROVED/)).toBeTruthy();
    expect(screen.getByText(/T/)).toBeTruthy();
    expect(screen.getByText('4/24/2019')).toBeTruthy();
  });

  it('rejected', () => {
    const rejection = {
      rejectedBy: {
        firstName: 'TJ',
        lastName: 'Jiang',
      },
      rejectedAt: '2019-4-24',
    };
    render(
      <IntlProvider locale="en">
        <TaskApprovalStatusInput rejection={rejection} />
      </IntlProvider>
    );

    expect(screen).toMatchSnapshot();
    expect(screen.getByText(/REJECTED/)).toBeTruthy();
  });
});
