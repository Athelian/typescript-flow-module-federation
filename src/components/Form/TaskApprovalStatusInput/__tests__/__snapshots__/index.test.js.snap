// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskApprovalStatusInput approved, showUser, showDate 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-1:hover > button {
  opacity: 1;
}

.emotion-4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-2 {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 12px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-transform: uppercase;
  text-align: center;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 0;
  border-radius: 999px;
  height: 40px;
  width: 200px;
  cursor: inherit;
  padding: 5px;
  background-color: #0b6ede;
  color: #ffffff;
}

.emotion-5:focus {
  outline: 0;
}

.emotion-5[disabled] {
  cursor: default;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #fff;
  width: 30px;
  height: 30px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  font-size: 14px;
  background-color: #ccc;
  box-shadow: 0 0 10px rgba(0,0,0,0.1);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<div
  class="emotion-5"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-0"
      tabindex="-1"
    >
      T
    </div>
  </div>
  <div
    class="emotion-4"
  >
    <div
      class="emotion-2"
    >
      APPROVED
    </div>
    <div
      class="emotion-2"
    >
      4/24/2019
    </div>
  </div>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-check-circle fa-w-16 fa-fw "
    data-icon="check-circle"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
      fill="currentColor"
    />
  </svg>
</div>
`;

exports[`TaskApprovalStatusInput rejected 1`] = `
.emotion-1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-0 {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 12px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-transform: uppercase;
  text-align: center;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 0;
  border-radius: 999px;
  height: 40px;
  width: 200px;
  cursor: inherit;
  padding: 5px;
  background-color: #ef4848;
  color: #ffffff;
}

.emotion-2:focus {
  outline: 0;
}

.emotion-2[disabled] {
  cursor: default;
}

<div
  class="emotion-2"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-0"
    >
      REJECTED
    </div>
  </div>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-times-circle fa-w-16 fa-fw "
    data-icon="times-circle"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
      fill="currentColor"
    />
  </svg>
</div>
`;
